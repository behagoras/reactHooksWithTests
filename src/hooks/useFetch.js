import { useEffect, useRef, useState } from 'react'

export default function useFetch(url, options = {}) {
  const isMounted = useRef(true)
  const [state, setState] = useState({ data: null, loading: true, error: null })
  const endError = (error) => isMounted.current && setState({ data: null, loading: false, error })
  const setLoading = (bool) => isMounted.current && setState({ ...state, loading: bool })

  useEffect(() => {
    return () => {
      isMounted.current = false
    }
  }, [])

  useEffect(() => {
    setLoading(true)
    fetch(url, options)
      .then((result) => result.json())
      .then((data) => isMounted.current && setState({ ...state, loading: false, data }))
      .catch(endError)
  }, [url])

  return state
}
