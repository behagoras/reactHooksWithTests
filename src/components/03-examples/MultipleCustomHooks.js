import React, { useEffect, useState } from 'react'
import useCounter from '../../hooks/useCounter'
import useFetch from '../../hooks/useFetch'

export default function MultipleCustomHooks() {
  const baseUrl = 'https://www.breakingbadapi.com/api'

  const [
    counter,
    increment,
    decrement,
    reset,
  ] = useCounter(1, 1)
  const [quoteUrl, setQuoteUrl] = useState(`${baseUrl}/quotes/${counter}`)
  const state = useFetch(quoteUrl)

  const { data = [], error, loading } = state

  useEffect(() => { setQuoteUrl(`${baseUrl}/quotes/${counter}`) }, [counter])

  return (
    <div>
      <h1>Breaking Bad characters</h1>
      <hr />
      {loading ? (
        <div className="alert alert-info text-center">
          Loading...
        </div>
      ) :
        data && data.map((quote, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <blockquote className="blockquote text-right" key={index}>
            <p className="mb-0">{quote?.quote}</p>
            <footer className="blockquote-footer">{quote?.author}</footer>
          </blockquote>
        ))}
      {/* {data && <pre>{JSON.stringify(data, 0, 1)}</pre>} */}
      <div className="buttons text-right">
        <button
          disabled={counter <= 1}
          type="button"
          className="btn btn-primary "
          onClick={counter > 1 && decrement}
        >
          {'<'}
        </button>
        <button
          type="button"
          className="btn "
          onClick={reset}
        >
          Reset
        </button>
        <button
          type="button"
          className="btn btn-primary"
          onClick={increment}
        >
          {'>'}
        </button>
      </div>
      {error && <pre>{JSON.stringify(error, 0, 1)}</pre>}
    </div>
  )
}

