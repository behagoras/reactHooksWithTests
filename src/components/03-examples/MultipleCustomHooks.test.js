import React from 'react'
import { shallow } from 'enzyme'
import MultipleCustomHooks from './MultipleCustomHooks'
import useFetch from '../../hooks/useFetch'

jest.mock('../../hooks/useFetch')

describe('<MultipleCustomHooks />', () => {
  it('should render correctly', () => {
    useFetch.mockReturnValue({
      loading: true,
      data: null,
      error: null,
    })
    const component = shallow(<MultipleCustomHooks />)
    expect(component).toMatchSnapshot()
  })
  it('should display the information', () => {
    const author = 'author'
    const quote = 'Hello World'
    useFetch.mockReturnValue({
      loading: false,
      data: [{ author, quote }],
      error: null,
    })
    const component = shallow(<MultipleCustomHooks />)

    expect(component.find('.mb-0').text().trim()).toBe(quote.trim())
    expect(component.find('footer').text().trim()).toBe(author.trim())
    expect(component.find('.alert').exists()).toBe(false)
  })
})
