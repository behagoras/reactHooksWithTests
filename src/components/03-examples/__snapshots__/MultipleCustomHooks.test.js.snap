// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleCustomHooks /> should render correctly 1`] = `
<div>
  <h1>
    Breaking Bad characters
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <div
    className="buttons text-right"
  >
    <button
      className="btn btn-primary "
      disabled={true}
      onClick={false}
      type="button"
    >
      &lt;
    </button>
    <button
      className="btn "
      onClick={[Function]}
      type="button"
    >
      Reset
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
  </div>
</div>
`;
